devito/tools/utils.py

```diff
@@ -0,0 +1,245 @@
+    i2 = type(iterable)(i for i in iterable if not f(i))
+    return i1, i2
+
+
+def roundm(x, y):
+    """Return x rounded up to the closest multiple of y."""
+    return x if x % y == 0 else x + y - x % y
+
+
+def powerset(iterable):
+    "powerset([1,2,3]) --> () (1,) (2,) (3,) (1,2) (1,3) (2,3) (1,2,3)"
+    s = list(iterable)
+    return chain.from_iterable(combinations(s, r) for r in range(len(s)+1))
+
+
+def invert(mapper):
+    """Invert a dict of lists preserving the order."""
+    inverse = OrderedDict()
+    for k, v in mapper.items():
+        for i in v:
+            inverse[i] = k
+    return inverse
+
+
+def flatten(l):
+    """Flatten a hierarchy of nested lists into a plain list."""
+    newlist = []
+    for el in l:
+        if isinstance(el, Iterable) and not isinstance(el, (str, bytes)):
+            for sub in flatten(el):
+                newlist.append(sub)
+        else:
+            newlist.append(el)
+    return newlist
+
+
+def single_or(l):
+    """Return True iff only one item is different than ``None``, False otherwise.
+    Note that this is not a XOR function, according to the truth table of the XOR
+    boolean function with n > 2 inputs. Hence the name ``single_or``."""
+    # No
+    i = iter(l)
+    return any(i) and not any(i)
+
+
+def filter_ordered(elements, key=None):
+    """Filter elements in a list while preserving order.
+
+    :param key: Optional conversion key used during equality comparison.
+    """
+    seen = set()
+    if key is None:
+        key = lambda x: x
+    return [e for e in elements if not (key(e) in seen or seen.add(key(e)))]
+
+
+def filter_sorted(elements, key=None):
+    """Filter elements in a list and sort them by key. The default key is
+    ``operator.attrgetter('name')``."""
+    if key is None:
+        key = attrgetter('name')
+    return sorted(filter_ordered(elements, key=key), key=key)
+
+
+def numpy_to_ctypes(dtype):
+    """Map numpy types to ctypes types."""
+    return {np.int32: ctypes.c_int,
+            np.float32: ctypes.c_float,
+            np.int64: ctypes.c_int64,
+            np.float64: ctypes.c_double}[dtype]
+
+
+def ctypes_to_C(ctype):
+    """Map ctypes types to C types."""
+    if issubclass(ctype, ctypes.Structure):
+        return 'struct %s' % ctype.__name__
+    elif issubclass(ctype, ctypes.Union):
+        return 'union %s' % ctype.__name__
+    elif ctype.__name__.startswith('c_'):
+        # FIXME: Is there a better way of extracting the C typename ?
+        # Here, we're following the ctypes convention that each basic type has
+        # the format c_X_p, where X is the C typename, for instance `int` or `float`.
+        return ctype.__name__[2:-2]
+    else:
+        raise TypeError('Unrecognised %s during converstion to C type' % str(ctype))
+
+
+def ctypes_pointer(name):
+    """Create a ctypes type representing a C pointer to a custom data type ``name``."""
+    return type("c_%s_p" % name, (ctypes.c_void_p,), {})
+
```

Comment:
shouldn't these three be in `types` or something?

Reply:
I **think** this is a lower level thing

Reply:
Yes it is, just seem a bit out of place there. I'll leave it to you to decide. GTG otherwise